data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/client/PeerForwarderClient.java

```diff
@@ -110,25 +124,37 @@ private WireEvent getWireEvent(final Event event) {
         );
     }
 
-    private CompletableFuture<AggregatedHttpResponse> processHttpRequest(final WebClient client, final String content) {
+    private CompletableFuture<AggregatedHttpResponse> processHttpRequest(final WebClient client, final String content,
+                                                                         final Collection<Record<Event>> records,
+                                                                         final PeerForwarderReceiveBuffer<Record<Event>> peerForwarderReceiveBuffer) {
         return CompletableFuture.supplyAsync(() ->
         {
-            final CompletableFuture<AggregatedHttpResponse> aggregate = client.post(DEFAULT_PEER_FORWARDING_URI, content).aggregate();
-            return aggregate.join();
-        }, executorService);
+            try {
+                final CompletableFuture<AggregatedHttpResponse> aggregate = client.post(DEFAULT_PEER_FORWARDING_URI, content).aggregate();
+                return aggregate.join();
+            } catch (final Exception e) {
+                processFailedRequestsLocally(null, records, peerForwarderReceiveBuffer);
+                throw e;
+            }
+        }, executorService).thenApply(httpResponse -> processFailedRequestsLocally(httpResponse, records, peerForwarderReceiveBuffer));
     }
 
-    private AggregatedHttpResponse getAggregatedHttpResponse(final CompletableFuture<AggregatedHttpResponse> aggregatedHttpResponseCompletableFuture) throws UnprocessedRequestException {
-        try {
-            return aggregatedHttpResponseCompletableFuture.get();
-        } catch (final InterruptedException e) {
-            LOG.error("Peer forwarding interrupted.");
-            throw new RuntimeException(e);
-        } catch (final ExecutionException e) {
-            if (e.getCause() instanceof RuntimeException)
-                throw (RuntimeException) e.getCause();
-            throw new RuntimeException(e.getCause());
+    private AggregatedHttpResponse processFailedRequestsLocally(final AggregatedHttpResponse httpResponse, final Collection<Record<Event>> records,
+                                                                final PeerForwarderReceiveBuffer<Record<Event>> peerForwarderReceiveBuffer) {
+        if (httpResponse == null || httpResponse.status() != HttpStatus.OK) {
+            try {
+                peerForwarderReceiveBuffer.writeAll(records, FAILED_REQUESTS_BUFFER_WRITE_TIMEOUT_MILLIS);
+            } catch (final Exception e) {
+                LOG.error("Unable to write failed records to local peer forwarder receive buffer due to exception. Dropping data.", e);
```

Comment:
It could be good to log how many records were dropped in this log line.

Reply:
Good call, will add